### DISCUSSION TITLE
Opening a dialog from a view model

### DISCUSSION THREAD

**Discussion by naavis** [AvaloniaUI/Avalonia#7924] [Q&A] [answered]
The Avalonia docs for dialogs refer to `Application.Current.MainWindow`, which I don't think exists in the current versions of Avalonia: https://docs.avaloniaui.net/docs/controls/window#show-a-window-as-a-dialog

What would be a proper way to reference the current root window? I'd like to show a file picker dialog from a view model command, but `OpenFileDialog.ShowAsync` requires a reference to a parent window. How do I obtain one if `Application.Current.MainWindow` does not exist?

I realize that opening file dialogs in the view model might not be the best way of doing things, though.

**Comment by naavis** [comment]
I was able to work around this by passing a reference to the parent window as a command parameter in XAML like this, but it feels hacky:
```
<MenuItem Header="_Open..." Command="{Binding OpenFile}" CommandParameter="{Binding $parent[Window]}"/>
```

**Comment by jeromemanzano** [comment]
Have you tried `Application.Current.ApplicationLifetime`? https://github.com/AvaloniaUI/Avalonia/issues/3418

**Reply by naavis** [reply]
Thanks, I'll try it out later today!

**Comment by robloo** [comment]
I'm also curious what the 'proper' way of doing things here is. In my opinon passing a window shouldn't even be required (it isn't in WPF). It should just use the default unless one is specified.

On my end, here is what I do. This probably isn't 'correct':

```csharp
    public class App : Application
    {
        /// <summary>
        /// Contains a reference to the main window of the application.
        /// </summary>
        public static MainWindow MainWindow;

        public override void Initialize()
        {
            AvaloniaXamlLoader.Load(this);
        }

        public override void OnFrameworkInitializationCompleted()
        {
            if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
            {
                desktop.MainWindow = new MainWindow();

                // Here is where a reference to the main window is saved to the App class
                App.MainWindow = desktop.MainWindow as MainWindow;
            }

            base.OnFrameworkInitializationCompleted();
        }
    }
```

This is a static reference (not guaranteeing that is a good idea for all apps) but it allows `App.MainWindow` to be used anywhere to get the reference.

**Reply by maxkatz6** [reply]
You can just show a window like in WPF, but if you want better integration with the system, you need to set ApplicationLifetime.MainWindow.
Simplest use case - depending on configuration application will be closed when MainWindow is closed. And it's not necessary a _first_ opened window. macOS has some special ways to work with windows, so we try to be friendly to their concept.

**Reply by robloo** [reply]
The code example above is really just to illustrate how to get a MainWindow reference and keep it available to all code. What I was trying to say is ShowDialog() itself shouldn't require a window. It just use the default focused one unless one is specified. Apologies for the poor wording above.

**Comment by naavis** [comment]
Like I mentioned in the starting message, opening a dialog in _the view model_ is probably a bad idea, because you want to be able to unit test the view model and so on. The music store app example in the Avalonia docs uses the ReactiveUI `Interaction` thingy to open a dialog in the code-behind file where it belongs, and then pass the result of the things done in the dialog on to the view model: https://docs.avaloniaui.net/tutorials/music-store-app/opening-a-dialog

I have not totally wrapped my head around it yet, but seems kinda like the "right way".

**Reply by maxkatz6** [reply]
Yes, window in a view model can be considered as a bad idea, because normally view model doesn't know about UI layer.
ReactiveUI.Interaction is one of the ways to do it, personally I prefer it in my small applications.
Many people prefer creating something like "IApplicationController" or "IWindowManager" (doesn't matter how you call it), which purpose is to manage windows and dialogs while being an abstraction over UI framework.

**Reply by 0x5ea000000** [reply]
Do you have any reference to implementation of "IWindowManager" thing?

**Reply by timunie** [reply]
I have two samples in preview mode: 

-https://github.com/timunie/Avalonia.Samples/tree/feature/MVVM_Dialogs/src/Avalonia.Samples/MVVM/MvvmDialogSample 

- https://github.com/timunie/Avalonia.Samples/tree/feature/MVVM_Dialogs/src/Avalonia.Samples/MVVM/AdvancedMvvmDialogSample

EDIT:
Here is the released version: https://github.com/AvaloniaUI/Avalonia.Samples?tab=readme-ov-file#-view-interaction-samples

**Reply by 0x5ea000000** [reply]
thank you very much with your Sample repo. I'm new to Desktop app development and choose Avalonia because it's a true cross platform solution around the dotnet desktop app war. But my very first impression was that it was quite difficult to use and docs is not complete. I think your repo will help me a lot.

**Reply by andreasgrosche** [reply]
@timunie could you please update the links? They lead to '404 - page not found'.

@naavis's solution has the benefit that a view model instance can be used by multiple views in different windows and the dialogs are shown with the correct owner. This is required, e.g., setting the dialog startup location to the center of the parent/owner.
I am curious whether this is also possible using the `Interaction` approach.

> ```
> <MenuItem Header="_Open..." Command="{Binding OpenFile}" CommandParameter="{Binding $parent[Window]}"/>
> ```

**Comment by maxkatz6** [comment, marked as answer]
Thanks for pointing on mistake in the docs. I updated the page to match new 0.10 APIs.

**Reply by naavis** [reply]
Thanks!
